mountpoint-s3/src/data_cache/multilevel_cache.rs

```diff
@@ -0,0 +1,314 @@
+
+use crate::object::ObjectId;
+
+use super::{BlockIndex, ChecksummedBytes, DataCache, DataCacheResult};
+
+pub struct MultilevelDataCache<DiskCache, ExpressCache, Runtime> {
+    disk_cache: Arc<DiskCache>,
+    express_cache: ExpressCache,
+    runtime: Runtime,
+}
+
+impl<DiskCache: DataCache, ExpressCache: DataCache, Runtime: Spawn>
+    MultilevelDataCache<DiskCache, ExpressCache, Runtime>
+{
+    pub fn new(disk_cache: Arc<DiskCache>, express_cache: ExpressCache, runtime: Runtime) -> Self {
+        assert_eq!(disk_cache.block_size(), express_cache.block_size());
+        Self {
+            disk_cache,
+            express_cache,
+            runtime,
+        }
+    }
+}
+
+#[async_trait]
+impl<DiskCache, ExpressCache, Runtime> DataCache for MultilevelDataCache<DiskCache, ExpressCache, Runtime>
+where
+    DiskCache: DataCache + Sync + Send + 'static,
+    ExpressCache: DataCache + Sync,
+    Runtime: Spawn + Sync,
+{
+    async fn get_block(
+        &self,
+        cache_key: &ObjectId,
+        block_idx: BlockIndex,
+        block_offset: u64,
+    ) -> DataCacheResult<Option<ChecksummedBytes>> {
+        match self.disk_cache.get_block(cache_key, block_idx, block_offset).await {
+            Ok(Some(data)) => {
+                trace!(cache_key=?cache_key, block_idx=block_idx, "block served from the disk cache");
+                return DataCacheResult::Ok(Some(data));
+            }
+            Ok(None) => (),
+            Err(err) => warn!(cache_key=?cache_key, block_idx=block_idx, ?err, "error reading block from disk cache"),
+        }
+
+        if let Some(data) = self.express_cache.get_block(cache_key, block_idx, block_offset).await? {
+            trace!(cache_key=?cache_key, block_idx=block_idx, "block served from the express cache");
+            let cache_key = cache_key.clone();
+            let cache = self.disk_cache.clone();
+            let data_cloned = data.clone();
+            self.runtime
+                .spawn(async move {
+                    if let Err(error) = cache
+                        .put_block(cache_key.clone(), block_idx, block_offset, data_cloned)
+                        .await
+                    {
+                        warn!(cache_key=?cache_key, block_idx, ?error, "failed to update the local cache");
+                    }
+                })
+                .unwrap();
+            return DataCacheResult::Ok(Some(data));
+        }
+
+        DataCacheResult::Ok(None)
+    }
+
+    async fn put_block(
+        &self,
+        cache_key: ObjectId,
+        block_idx: BlockIndex,
+        block_offset: u64,
+        bytes: ChecksummedBytes,
+    ) -> DataCacheResult<()> {
+        match self
+            .disk_cache
+            .put_block(cache_key.clone(), block_idx, block_offset, bytes.clone())
+            .await
+        {
+            Ok(_) => (),
+            Err(error) => warn!(cache_key=?cache_key, block_idx, ?error, "failed to update the local cache"),
```

Comment:
Stylistically, do we prefer this or using something like `.inspect_err` or `if let Err(...) = ...` for stuff like this?

Reply:
`if let Err(...) = ...` looks better, I'll use it here